docs/hd.md

```diff
@@ -1,124 +1,191 @@
-you can have different accounts for different purposes and so on.
+## HD wallet
 
-## Derivation process
+HD (Hierarchical Deterministic) wallet has hierarchical, tree-based structure.
+The root of this tree is a pair of user's secret key `RootSK` and corresponding
+public key `RootPK`. `RootSK` is a key we obtain during creation of the wallet:
+by default `RootSK` is generating from 12-words mnemonic (or "wallet backup
+phrase").
 
-To derive secret key, parent secret key and passphrase are necessary.
-Derivation is done using `deriveHDSecretKey`.
+`RootSK` and `RootPK` are crucially important keys:
 
-The deriving of public key can be performed using just a public key of a parent node.
-Derivation is done using `deriveHDPublicKey`.
+* Wallet's identifier is based on `RootPK`.
+* `RootSK` is using to sign transactions when we want to spend money from this
+  wallet.
+* `RootSK` is using to derive new addresses for this wallet (see below).
+* `RootPK` is using during synchronization with the blockchain to find out
+  current balance of this wallet.
 
-It's not allowed to derive a public key from parent public key for
-indices which are greater than or equal to 2³¹:
-for these indicies parent secret key is required.
-This feature is called "hardened derivation":
+So, who owns the root keys - he owns the wallet and all its money.
 
-You can get description of underlying cryptography [here](https://cardanolaunch.com/assets/Ed25519_BIP.pdf).
+## Tree structure
 
-Note: all these functions don't require derivation level number,
-only credential information (public key or secret key with passphrase)
-and index of a child.
-Also we can derive public and secret keys for any arbitrary number of levels.
+In general case structure of the tree can be arbitrary, but we use two layered
+structure: nodes of the first layer are called "accounts", nodes of the second
+layer are called "addresses" (you can think of them as leaves of the tree).
+Sometimes layer is called "level". It can be represented like this:
+
+              first            second
+              layer/level      layer/level
+
+    Root --
+           `- Account [0] --
+           `                `- Address [0]
+           `                `- Address [1]
+           `                `  ...
+           `                `- Address [M]
+           `
+           `- Account [1] --
+           `                `- ...
+           `  ...
+           `- Account [N] --
+                            `- ...
+
+There can be potentially 2³² accounts (`N` is `Word32` value) and each account
+can have 2³² addresses (`M` is `Word32` value as well), so wallet can have 2⁶⁴
+addresses in total.
+
+So each address belongs to one account. Address can contain money, be used as
+a change address and so on.
+
+Balance of each account can be computed as a sum of balances of addresses which
+belong to it, and therefore balance of the wallet is a sum of balances of its
+accounts.
```

Comment:
Don't think you need to describe all the subtleties here, but it would be useful to refer to the wallet formal specification here, because if somebody takes this description to implement balance computation it will most likely be incorrect.